### PR TITLE
Add allowed-tenant config to allow cross-tenant domain access for whitelisted tenants

### PR DESCRIPTION
**Pull request by piraveena**
### Proposed changes in this pull request

This will allow certain some white-listed tenants to perform cross-tenant access.  `allowed_tenants="carbon.super"` config needs to be added to deployment.toml file where it can have comma separated tenants.  Related PR https://github.com/wso2/carbon-identity-framework/pull/3784

```
[[resource.access_control]]
context="(.*)/oauth2/introspect(.*)"
secure = "true"
http_method = "all"
cross_tenant = true
cross_access_allowed_tenants="carbon.super"
```

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/module/ResourceConfig.java [hunk 1]
```diff
@@ -37,6 +37,7 @@
     private List<String> scopes;
     // Comma separated list of allowed authentication handler names. If all handlers are engaged the value is 'all'
     private String allowedAuthHandlers;
+    private String allowedTenants;
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/module/ResourceConfig.java:40]: crossAcessAllowedTenants

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/util/Utils.java [hunk 2]
```diff
@@ -66,4 +69,31 @@ public static boolean isUserBelongsToRequestedTenant(AuthenticationContext authe
         }
         return tenantDomainFromURLMapping.equals(tenantDomain);
     }
+
+    /**
+     * Checks whether cross-tenant domain is allowed for the tenant.
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/util/Utils.java:74]: `WDYT of something like this Check whether the tenant domain is allowed as a cross-tenant domain`?

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/util/Utils.java [hunk 3]
```diff
@@ -66,4 +69,31 @@ public static boolean isUserBelongsToRequestedTenant(AuthenticationContext authe
...
+     *
+     * @param authenticationContext Context of the authentication
+     * @param authorizationContext  Context of the authorization
+     * @return true if the tenant is allowed or if the config is null.
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/util/Utils.java:78]: WDYT `True if....`?

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java [hunk 4]
```diff
@@ -120,9 +120,17 @@ public void buildResourceAccessControlData() {
...
-                            Boolean.FALSE.toString().equals(isCrossTenantAllowed))) {
+                    String crossAccessAllowedTenants =
+                            resource.getAttributeValue(new QName(Constants.RESOURCE_CROSS_ACCESS_ALLOWED_TENANTS));
+
+                    if (StringUtils.isNotEmpty(isCrossTenantAllowed) &&
+                            (Boolean.TRUE.toString().equals(isCrossTenantAllowed)
+                                    || Boolean.FALSE.toString().equals(isCrossTenantAllowed))) {
                         resourceConfig.setIsCrossTenantAllowed(Boolean.parseBoolean(isCrossTenantAllowed));
+                        if (resourceConfig.isCrossTenantAllowed() && StringUtils.isNotEmpty(crossAccessAllowedTenants))
+                        {
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java:131]: Can we move the `{` to the above line?

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java [hunk 5]
```diff
@@ -229,6 +237,24 @@ public void buildIntermediateCertValidationConfigData() {
...
+        if (StringUtils.isNotBlank(crossAccessAllowedTenants)) {
+            List<String> allowedTenantDomainsList = new ArrayList<>();
+            String regex = "\\s*,\\s*";
+            String[] allowedTenantDomainsNames = crossAccessAllowedTenants.split(regex);
+            allowedTenantDomainsList.addAll(Arrays.asList(allowedTenantDomainsNames));
+            return allowedTenantDomainsList;
+        }
+        return null;
+    }
+
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java:257]: This has only a single usage, right? If that's the case do we need to provide this as a public method? Also, we can pass any string with `,` and get a list. So we cannot say the returned list is a tenant list. WDYT?

Comment (somindatommy) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/util/AuthConfigurationUtil.java:257]: I don't think this is something to be given as public util method

**Review by IsuraD — APPROVED** [review]

**Pull request merged**